File: rust/src/benchmarks/json_parse_mapping.rs
use super::super::{helper, Benchmark};
use crate::config_i64;
use serde::{Deserialize, Serialize};

#[derive(Deserialize, Serialize)]
struct CoordinateData {
    x: f64,
    y: f64,
    z: f64,
}

#[derive(Deserialize)]
struct Coordinates {
    coordinates: Vec<CoordinateData>,
}

pub struct JsonParseMapping {
    n: i64,
    text: String,
    result_val: u32,
}

impl JsonParseMapping {
    pub fn new() -> Self {
        let n = config_i64("JsonParseMapping", "coords");

        Self {
            n,
            text: String::new(),
            result_val: 0,
        }
    }

    fn calc(&self, text: &str) -> CoordinateData {
        let data: Coordinates = serde_json::from_str(text).unwrap();
        let len = data.coordinates.len() as f64;

        let mut x = 0.0;
        let mut y = 0.0;
        let mut z = 0.0;

        for coord in data.coordinates {
            x += coord.x;
            y += coord.y;
            z += coord.z;
        }

        CoordinateData {
            x: x / len,
            y: y / len,
            z: z / len,
        }
    }
}

impl Benchmark for JsonParseMapping {
    fn name(&self) -> String {
        "JsonParseMapping".to_string()
    }

    fn prepare(&mut self) {
        use crate::benchmarks::json_generate::JsonGenerate;

        let mut generator = JsonGenerate::new();
        generator.n = self.n;
        generator.prepare();
        generator.run(0);
        self.text = generator.get_result().to_string();
    }

    fn run(&mut self, _iteration_id: i64) {
        let avg = self.calc(&self.text);
        self.result_val = self.result_val.wrapping_add(
            helper::checksum_f64(avg.x)
                .wrapping_add(helper::checksum_f64(avg.y))
                .wrapping_add(helper::checksum_f64(avg.z)),
        );
    }

    fn checksum(&self) -> u32 {
        self.result_val
    }
}

File: kotlin/src/main/kotlin/benchmarks/JsonParseMapping.kt
package benchmarks

import Benchmark
import com.alibaba.fastjson2.*

data class Coord(val x: Double, val y: Double, val z: Double)

class JsonParseMapping : Benchmark() {
    private lateinit var text: String
    private var resultVal: UInt = 0u

    override fun prepare() {
        val generator = JsonGenerate()
        generator.n = configVal("coords")
        generator.prepare()
        generator.run(0)
        text = generator.text
    }

    private fun calc(text: String): Coord {

        val data = text.into<CoordinatesData>()
        val coordinates = data.coordinates

        var x = 0.0
        var y = 0.0
        var z = 0.0

        for (coord in coordinates) {
            x += coord.x
            y += coord.y
            z += coord.z
        }

        val len = coordinates.size.toDouble()
        return Coord(x / len, y / len, z / len)
    }

    class CoordinatesData {
        var coordinates: List<Coord> = emptyList()
    }

    override fun run(iterationId: Int) {
        val coord = calc(text)
        resultVal += Helper.checksum("%.7f".format(coord.x)) +
                    Helper.checksum("%.7f".format(coord.y)) +
                    Helper.checksum("%.7f".format(coord.z))  
    }

    override fun checksum(): UInt = resultVal

    override fun name(): String = "JsonParseMapping"
}